### ISSUE TITLE
Integrate HF peft for LoRA/adapter support (replaces custom LoRALinear)

### ISSUE THREAD

**Issue by timlawrenz** [timlawrenz/DINO-X#19]
## Summary
Integrate HuggingFace `peft` library for parameter-efficient fine-tuning instead of building a custom LoRALinear module. This gives us LoRA, QLoRA, adapter merging, and HF Hub compatibility for free.

## Revised Design

Use `peft.LoraConfig` + `get_peft_model()` to inject low-rank adapters into our `PatchViT`:

```python
from peft import LoraConfig, get_peft_model

config = LoraConfig(
    r=8,
    lora_alpha=16,
    target_modules=["qkv", "proj", "fc1", "fc2"],
    lora_dropout=0.05,
)
model = get_peft_model(backbone, config)
# Only ~200K params trainable, backbone frozen
```

Provide a thin wrapper in `zoo/peft.py` for the clinical researcher workflow:

```python
from zoo.peft import apply_lora, load_adapter

model = PatchViT(...)
model = apply_lora(model, rank=8)
# ... train on local hospital data ...
model.save_pretrained("dinox-ct-pe-adapter")
```

## Expected Parameter Sizes (unchanged)
| Rank | Params (QKV+MLP, ViT-S) | Size |
|------|-------------------------|------|
| 8 | ~200K | ~0.8MB |
| 32 | ~800K | ~3MB |
| 64 | ~1.6M | ~6MB |

vs backbone: 24.9M params, ~95MB

## Acceptance Criteria
- [ ] `peft` added to `requirements.in` (optional dep for fine-tuning)
- [ ] `PatchViT` linear layers discoverable by `peft` (`target_modules` works)
- [ ] ScaleEmbedding passthrough works correctly with frozen backbone + LoRA
- [ ] Only adapter params have `requires_grad=True`
- [ ] `save_pretrained()` / `from_pretrained()` round-trip works
- [ ] Thin wrapper in `zoo/peft.py` with `apply_lora()` and `load_adapter()`
- [ ] Unit tests for gradient isolation, save/load round-trip, scale embedding compat

## Depends On
- #2 (ViT with ScaleEmbedding — need the final architecture)

## Why not custom LoRALinear?
See [comment](https://github.com/timlawrenz/DINO-X/issues/19#issuecomment-4288683777) — rolling our own LoRA introduces technical debt, breaks HF ecosystem compatibility, and duplicates battle-tested code.

**Comment by timlawrenz** [comment]
## Revised Approach: Use HF `peft` Instead of Custom LoRALinear

The original plan to build a custom `LoRALinear` drop-in replacement is the wrong path. HuggingFace's [`peft`](https://github.com/huggingface/peft) library is the industry standard for parameter-efficient fine-tuning and already provides everything we need.

### Why `peft` over custom implementation

| Concern | Custom LoRALinear | HF `peft` |
|---------|-------------------|-----------|
| LoRA injection | Manual module replacement | `get_peft_model(model, LoraConfig(...))` — one line |
| Merge/unmerge | Must implement ourselves | Built-in `merge_adapter()` / `unmerge_adapter()` |
| Serialization | Custom save/load | `save_pretrained()` / `from_pretrained()` — HF Hub compatible |
| Adapter composition | Not supported | Adapter merging, stacking, weighted combination |
| Quantization compat | Not supported | QLoRA (4-bit/8-bit) out of the box |
| Community sharing | Custom format | Push adapters to HF Hub, any researcher can `PeftModel.from_pretrained()` |
| Maintenance burden | Ours forever | Maintained by HF team + community |

### Revised plan

1. **Make `PatchViT` peft-compatible** — Ensure `peft` can identify the `nn.Linear` layers in our ViT (attention QKV projections, MLP layers). This may require minor naming adjustments.

2. **Add a thin integration wrapper** in `zoo/` that provides a clean API:
   ```python
   from zoo.peft import apply_lora, load_adapter
   
   model = PatchViT(...)  # load pretrained backbone
   model = apply_lora(model, rank=8, target_modules=["qkv", "proj", "fc1", "fc2"])
   # Train adapter...
   model.save_pretrained("dinox-ct-vit-small-pe-adapter")
   ```

3. **Document the clinical researcher workflow** — Pull backbone, freeze it, train a 5MB adapter on local data, share on HF Hub.

4. **Add `peft` to `requirements.in`** as an optional dependency (only needed for fine-tuning, not inference).

### What stays the same
- Acceptance criteria for gradient isolation (only adapter params have `requires_grad=True`)
- Expected parameter sizes (rank 8 → ~0.8MB, rank 32 → ~3MB)
- The ScaleEmbedding passthrough must work correctly with frozen backbone + LoRA

### What gets dropped
- Custom `LoRALinear` class
- Custom `merge()` implementation
- Custom serialization

**Comment by timlawrenz** [comment]
## Compatibility Note: `nn.MultiheadAttention` vs `peft`

After inspecting the current `PatchViT` architecture, there's a concrete blocker to address:

**Current `TransformerBlock` uses `nn.MultiheadAttention`**, which stores QKV as a fused `in_proj_weight` parameter — not as separate `nn.Linear` submodules:

```
attn.in_proj_weight: [1152, 384]  ← fused QKV, raw Parameter
attn.out_proj: NonDynamicallyQuantizableLinear [384, 384]  ← this one works
```

`peft` can only inject LoRA into `nn.Linear` (or `nn.Conv2d`) *modules*, not raw parameters. So `target_modules=["qkv"]` won't find anything.

### What works today
- `blocks.*.mlp.0` (fc1) ✅ — plain `nn.Linear`
- `blocks.*.mlp.2` (fc2) ✅ — plain `nn.Linear`  
- `blocks.*.attn.out_proj` ✅ — is a `Linear` submodule

### What doesn't work
- `blocks.*.attn.in_proj_weight` ❌ — fused QKV as raw parameter, invisible to `peft`

### Fix: Refactor attention to use explicit Linear layers

Replace `nn.MultiheadAttention` with a custom attention block using separate `nn.Linear` layers, matching the `timm` ViT convention:

```python
class Attention(nn.Module):
    def __init__(self, dim, heads):
        super().__init__()
        self.qkv = nn.Linear(dim, dim * 3, bias=True)  # peft can target this
        self.proj = nn.Linear(dim, dim, bias=True)       # peft can target this
        self.heads = heads
        self.scale = (dim // heads) ** -0.5
```

This is a prerequisite for this issue. The refactor should be output-equivalent to the current `nn.MultiheadAttention` (verifiable with a checkpoint compatibility test).

### Freeze strategy for `apply_lora()`

`apply_lora()` should automatically freeze everything *except* LoRA params:
- `patch_embed` — frozen (convolution weights are pretrained)
- `pos_embed` — frozen
- `scale_embed` — **always frozen** (clinical physics must not drift during adaptation)
- `blocks.*.norm*` — frozen (or optionally trainable, configurable)
- LoRA A/B matrices — trainable

This means a researcher calling `apply_lora(model, rank=8)` gets safe defaults without needing to understand the internal architecture.

**Comment by timlawrenz** [comment]
## Critical: Use `F.scaled_dot_product_attention` After Refactor

When refactoring from `nn.MultiheadAttention` to explicit `nn.Linear` QKV layers, there is a **major performance trap**: PyTorch's `nn.MultiheadAttention` automatically routes to FlashAttention/memory-efficient SDPA kernels. A naive `q @ k.T / sqrt(d)` implementation will instantly OOM on high-resolution CT patch grids.

### Required pattern

```python
import torch.nn.functional as F

class Attention(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=True):
        super().__init__()
        self.num_heads = num_heads
        self.scale = (dim // num_heads) ** -0.5

        # Explicit Linear layers — peft can target these
        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.proj = nn.Linear(dim, dim)

    def forward(self, x):
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
        q, k, v = qkv.unbind(0)

        # CRITICAL: Route through SDPA — uses FlashAttention/mem-efficient kernels
        x = F.scaled_dot_product_attention(q, k, v)

        x = x.transpose(1, 2).reshape(B, N, C)
        x = self.proj(x)
        return x
```

### Why this matters for DINO-X specifically

- **ROCm + Strix Halo**: We rely on `TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1` for Triton-backed FlashAttention 2. `F.scaled_dot_product_attention` is the correct entry point for this — it dispatches to the AOTriton backend automatically.
- **RTX 4090 (24GB)**: At 512×512 input with patch=14, we get 1369 patches. Naive attention would allocate a `[B, heads, 1369, 1369]` score matrix per layer — ~28MB per sample at fp16. SDPA avoids materializing this.
- **Multi-day training stability**: SDPA handles the memory allocation internally with proven, tested code. Custom attention is a regression risk for our "no OOM crashes" constraint.

### Verification checklist for the refactor

1. Output equivalence: Load existing `nn.MultiheadAttention` checkpoint weights into new `Attention` module, verify identical forward pass output (within fp16 tolerance)
2. Memory: Profile peak VRAM before/after refactor at batch_size=64, confirm no regression  
3. SDPA dispatch: Verify FlashAttention kernel is actually selected (use `torch.backends.cuda.sdp_kernel()` context manager in tests)
4. `peft` targeting: Confirm `LoraConfig(target_modules=["qkv", "proj"])` finds all attention layers

**Comment by timlawrenz** [comment]
Implemented in 0b32675. zoo/peft.py provides apply_lora() with clinical safety guardrails: hard-freezes ScaleEmbedding, patch_embed, pos_embed so downstream adapters learn pathology features, not physics. Supports peft LoraConfig with configurable rank and target modules.
